src/Hilke.KineticConvolution/ConvolutionFactory.cs

```diff
@@ -157,20 +157,46 @@ public Shape<TAlgebraicNumber> CreateShape(IEnumerable<Tracing<TAlgebraicNumber>
             }
 
-            if (!isG1Continuous(tracingsEnumerated))
+            if (!verifyConditionOnConsecutive(tracingsEnumerated, continuityCondition, out var indicesWhereNotContinuous))
             {
-                throw new ArgumentException("The tracings should be continuous.", nameof(tracings));
+                throw new ArgumentException(
+                    $"The tracings must be continuous. Discontinuity detected at indices "
+                  + $"{string.Join(", ", indicesWhereNotContinuous.Select(index => $"{index} -> {nextIndex(index)}"))}",
+                    nameof(tracings));
+            }
+
+            if (!verifyConditionOnConsecutive(tracingsEnumerated, tangentContinuityCondition, out var indicesWhereTangentNotContinuous))
+            {
+                throw new ArgumentException(
+                    $"The tracings tangents must be continuous. Tangents discontinuity detected at indices "
+                  + $"{string.Join(", ", indicesWhereTangentNotContinuous.Select(index => $"{index} -> {nextIndex(index)}"))}",
+                    nameof(tracings));
             }
 
             return new Shape<TAlgebraicNumber>(tracingsEnumerated);
 
-            static bool isG1Continuous(IReadOnlyList<Tracing<TAlgebraicNumber>> tracings)
+            static bool continuityCondition(Tracing<TAlgebraicNumber> current, Tracing<TAlgebraicNumber> next) =>
+                current.IsContinuousWith(next);
+
+            static bool tangentContinuityCondition(Tracing<TAlgebraicNumber> current, Tracing<TAlgebraicNumber> next) =>
+                current.IsTangentContinuousWith(next);
+
+            static bool verifyConditionOnConsecutive(IReadOnlyList<Tracing<TAlgebraicNumber>> tracings,
+                                    Func<Tracing<TAlgebraicNumber>, Tracing<TAlgebraicNumber>, bool> condition,
+                                     out IReadOnlyList<int> indicesWhereConditionFails)
             {
-                return tracings.Zip(
-                                   tracings.Skip(1).Append(tracings.First()),
-                                   (right, left) => right.IsG1ContinuousWith(left))
-                               .All(isContinuous => isContinuous);
+                indicesWhereConditionFails = tracings.Zip(
+                                                         tracings.Skip(1).Append(tracings.First()),
+                                                         (current, next) => condition(current, next))
+                                                     .Select((satisfiesCondition, index) => (satisfiesCondition, index))
+                                                     .Where(e => !e.satisfiesCondition)
+                                                     .Select(e => e.index)
+                                                     .ToList();
+
+                return indicesWhereConditionFails.Count == 0;
```

Comment:
It is bad to simulate `(Cyclic)Pairwise` with `enumerable.Zip(enumerable.Skip(1), ...)` since it leads to a double enumeration. It would be better to implement (as `internal`) `CyclicPairwise` such as:
```csharp
        public static IEnumerable<TSelected> CyclicPairwise<TElement, TSelected>(
            this IEnumerable<TElement> source,
            Func<TElement, TElement, TSelected> selector)
        {
            if (source is null)
            {
                throw new ArgumentNullException(nameof(source));
            }

            if (selector is null)
            {
                throw new ArgumentNullException(nameof(selector));
            }

            return CyclicPairwiseIterator(); IEnumerable<TSelected> CyclicPairwiseIterator()
            {
                using var enumerator = source.GetEnumerator();

                if (!enumerator.MoveNext())
                {
                    yield break;
                }

                var firstElement = enumerator.Current;
                var previousElement = enumerator.Current;
                var closeCycle = false;

                while (enumerator.MoveNext())
                {
                    closeCycle = true;
                    var nextElement = enumerator.Current;

                    yield return selector(previousElement, nextElement);

                    previousElement = nextElement;
                }

                if (closeCycle)
                {
                    yield return selector(previousElement, firstElement);
                }
            }
        }
```